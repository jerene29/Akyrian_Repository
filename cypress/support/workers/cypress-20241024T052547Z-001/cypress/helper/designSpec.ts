import { GetStudyRevisionListDocument } from '../../src/graphQL/generated/graphql';
import { mockUserDataAdmin } from '../../src/constant/testFixtures';

type SpecNumber = `S-${number}-${number}`;

type SpecInfo = {
  // e.g S-09-010
  spec: SpecNumber;
  // title e.g headerSection, combined description becomes S-09-010 headerSection, but optional
  title?: string;
};

type Options = {
  loginAs?: string;
  skipTest?: boolean;
};

function resolveDescription(description: SpecNumber | SpecInfo) {
  return typeof description === 'string' ? description : `${description.spec} ${description.title}`;
}

const aliasGetStudyRevisionListDocument =
  'name' in GetStudyRevisionListDocument.definitions[0]
    ? GetStudyRevisionListDocument.definitions[0].name?.value
    : 'GetStudyRevisionListDocument';
/*
example usage
designSpec('S-19-2', () => {

  describe('...', ...)

  it('checks something...', () => {...})
});
*/
export function designSpec(
  description: SpecNumber | SpecInfo,
  callback: () => void,
  options?: Options,
) {
  before(() => {
    cy.intercept('POST', '/graphql', (req) => {
      if (req.body.operationName === aliasGetStudyRevisionListDocument) {
        req.alias = req.body.operationName;
      }
    });
    cy.beforeSetup(mockUserDataAdmin);
    cy.clearLocalStorageSnapshot();
    cy.saveLocalStorage();
    cy.wait(`@${aliasGetStudyRevisionListDocument}`);
  });
  if (options?.skipTest) {
    describe.skip(resolveDescription(description), callback);
  } else if (typeof description !== 'string' && description.title) {
    describe(description.spec, () => {
      describe(String(description.title), callback);
    });
  } else {
    describe(resolveDescription(description), callback);
  }
}
