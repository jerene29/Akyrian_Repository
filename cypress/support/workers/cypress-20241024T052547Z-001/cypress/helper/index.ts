export * from './designSpec';
export * from './object';