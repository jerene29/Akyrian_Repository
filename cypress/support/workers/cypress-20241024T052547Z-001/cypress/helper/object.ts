/**
 * to shorten the tedious '[data-cy=selector]' selector.
 * just put the name of data-cy attributed element.
 * -- d stands for document
 *
```
d`addButton`
```
 */
export function d(args: TemplateStringsArray) {
  return `[data-cy=${args.join('')}]`;
}
