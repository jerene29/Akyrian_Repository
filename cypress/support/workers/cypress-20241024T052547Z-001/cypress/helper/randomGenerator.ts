import moment = require("moment");

export const randomAlphaNumeric = (length: number) => {
  const chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
  let result = '';
  for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
  return result;
}

export const randomAlphabet = (length: number) => {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
  let result = '';
  for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
  return result;
}

export const randomDate = () => {
  const start = new Date(2012, 0, 1)
  const end = new Date()
  const date = new Date(start.getTime() + Math.random() * (end.getTime() - start.getTime()))
  return moment(date);
}