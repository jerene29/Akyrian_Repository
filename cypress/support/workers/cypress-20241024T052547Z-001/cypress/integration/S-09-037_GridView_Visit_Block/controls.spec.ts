import { d, designSpec } from '../../helper';

const options = {
  loginAs: 'admin@example.com',
};

designSpec(
  {
    spec: 'S-09-037',
    title: 'Control Buttons on Visit Block',
  },
  () => {
    before(() => {
      cy.createStudy({
        name: 'GRIDVIEW_TEST',
        formCount: 5,
        visitCount: 5,
      });
    });

    context('button done', () => {
      it('clickable on valid entry', () => {
        cy.get('body').type('{esc}');

        cy.hover(d`headerCell-0-3`);
        cy.get(d`visitNameBox`).click();
        cy.get(d`visitNameInput`)
          .focus()
          .clear()
          .type('Hello World');

        cy.get(d`editConfirm`).click();

        cy.hover(d`headerCell-0-2`);
        cy.hover(d`headerCell-0-3`);
      });

      it("invalid entry won't confirm the edit", () => {
        cy.get('body').type('{esc}');

        cy.hover(d`headerCell-0-4`);
        cy.get(d`visitNameBox`).click();
        cy.get(d`visitNameInput`)
          .focus()
          .clear()
          .type('...()***');

        cy.get(d`editConfirm`).click();

        cy.getTextSnapshot(d`cellErrorPopover`);
      });
    });

    context('button cancel', () => {
      it('checks if system behavior is the same as [ESC] key', () => {
        cy.get('body').type('{esc}');

        cy.hover(d`headerCell-0-3`);
        cy.get(d`visitNameBox`).click();
        cy.get(d`visitNameInput`)
          .focus()
          .clear()
          .type('Should Cancel');

        cy.get(d`editCancel`).click();

        cy.hover(d`headerCell-0-2`);
        cy.hover(d`headerCell-0-3`);

        // now we test ending the input with type {esc}
        cy.get('body').type('{esc}');

        cy.hover(d`headerCell-0-3`);
        cy.get(d`visitNameBox`).click();
        cy.get(d`visitNameInput`)
          .focus()
          .clear()
          .type('Should Cancel{esc}');

        cy.hover(d`headerCell-0-2`);
        cy.hover(d`headerCell-0-3`);
      });
    });

    context('tab sequence', () => {
      it('navigates through editable visit input while confirming the valid edit', () => {
        const EDITED_NAME = 'Edited Visit';
        cy.get('body').type('{esc}');

        cy.hover(d`headerCell-0-4`);

        cy.get('body').tab().tab();
        cy.shouldBeVisible(d`editableVisitWindowBefore-0-4`);
        cy.get('body').tab();
        cy.shouldBeVisible(d`editableVisitWindowAfter-0-4`);
        cy.get('body').tab();
        cy.shouldBeVisible(d`visitNameInput`)
          .focus()
          .clear()
          .type(EDITED_NAME);
        cy.get('body').tab();
        cy.shouldBeVisible(d`dayOffsetInput`);
        cy.get('body').tab();

        // the visit name should be `Edited Name`
        cy.get(d`headerCell-0-4`).contains(EDITED_NAME);
        cy.hover(d`headerCell-0-4`);
      });

      it('navigates backward to previous visit', () => {
        cy.get('body').type('{esc}');

        cy.get('body').type('{esc}');

        cy.hover(d`headerCell-0-4`);

        cy.get('body').tab({ shift: true });
        cy.shouldBeVisible(d`dayOffsetInput`)
          .clear()
          .type('9');

        cy.get('body').tab({ shift: true });
        cy.shouldBeVisible(d`visitNameInput`);
        cy.get('body').tab({ shift: true });
        cy.shouldBeVisible(d`editableVisitWindowAfter-0-3`);
        cy.get('body').tab({ shift: true });
        cy.shouldBeVisible(d`editableVisitWindowBefore-0-3`);
        cy.get('body').tab({ shift: true });
        cy.shouldBeVisible(d`relativeDayOffsetInput-0-3`);
        cy.get('body').tab({ shift: true });
        cy.get(d`relativeDayOffsetInput-0-3`).should('not.exist');
      });
    });

    context('snapshots text changes', () => {
      it('snapshot headerCells', () => {
        cy.reload();

        cy.intercept('POST', '/graphql', (req) => {
          req.alias = req.body.operationName;
        });

        cy.wait('@GetVisitTemplateList').then(() => {
          cy.get(d`grid-view`).click();
          cy.get(d`grid-view`).click();
          cy.get(d`grid-view`).click();

          cy.get('body').type('{esc}{esc}');
          cy.getTextSnapshot(d`headerCell-0-1`);
          cy.getTextSnapshot(d`headerCell-0-2`);
          cy.getTextSnapshot(d`headerCell-0-3`);
          cy.getTextSnapshot(d`headerCell-0-4`);
        });
      });
    });
  },
  options,
);
