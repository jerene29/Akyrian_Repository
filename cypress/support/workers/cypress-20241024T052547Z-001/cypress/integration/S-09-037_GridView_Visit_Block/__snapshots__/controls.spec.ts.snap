exports[`S-09-037 > Control Buttons on Visit Block > button done > invalid entry won't confirm the edit #0`] =
{
  "snapshot": "can only contain letters, numbers and special characters: [space], -, _, #"
};

exports[`S-09-037 > Control Buttons on Visit Block > button done > invalid entry won't confirm the edit #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Control Buttons on Visit Block > snapshots text changes > snapshot headerCells #0`] =
{
  "snapshot": "new visit 1Anchor Visit"
};

exports[`S-09-037 > Control Buttons on Visit Block > snapshots text changes > snapshot headerCells #1`] =
{
  "snapshot": "-2+2new visit 21 Day after Anchor"
};

exports[`S-09-037 > Control Buttons on Visit Block > snapshots text changes > snapshot headerCells #2`] =
{
  "snapshot": "-2+2Hello World9 Days after Anchor"
};

exports[`S-09-037 > Control Buttons on Visit Block > snapshots text changes > snapshot headerCells #3`] =
{
  "snapshot": "-2+2Edited Visit10 Days after Anchor"
};
